### DISCUSSION TITLE
Lemmatizer problem migrating from spaCy v2

### DISCUSSION THREAD

**Discussion by joaneeet7** [explosion/spaCy#9512] [Help: Coding & Implementations] [answered]
I have been using Spacy with the version 2.1.3. Therefore, I would like to update to the current version. However, I have the next problem:

`[E1004] Missing lemmatizer table(s) found for lemmatizer mode 'rule'. Required tables: ['lemma_rules']. Found: []. Maybe you forgot to call `nlp.initialize()` to load in the data?`

I have tried:
```
config={"model": None, "mode": "rule", "overwrite": False}
self.lemmatizer = self.nlp.add_pipe("lemmatizer", config=config)
self.lemmatizer.initialize()
```

But I'm not able to create de lemmatizer.

With the old version of Spacy we did it in the following way:
`**Word** -> Fails, **Lemma** -> Fail (**Version 2.3.1: with token.lemma_**)`

What is the way in Spacy v3?

**Comment by polm** [comment, marked as answer]
Sorry to hear you're having trouble with this. 

Can you clarify what you did that caused the E1004 to show up? Were you trying to load an old model, or trying to make a model in v3 with the same structure, or something else?

Note that at some point after 2.1 the lemmatizer tables were moved to a package called `spacy-lookups-data`, so you'll need to install that if you're making a Lemmatizer directly and want the lookup tables. Normally that's not necessary if you're using a pretrained pipeline.

> With the old version of Spacy we did it in the following way:
> `**Word** -> Fails, **Lemma** -> Fail (**Version 2.3.1: with token.lemma_**)`

Sorry, but I don't understand what you mean here. The way the `lemma_` attribute works is the same in v3 as it was in v2. 

-----

Adding a lemmatizer like the one in the pretrained pipelines to a custom pipeline is slightly complex, and the process differs depending on the language. Generally you need POS tags from either a Morphologizer or a Tagger + Attribute Ruler (see #8341). 

Can you clarify what language you're using and what your overall goal is?

**Reply by joaneeet7** [reply]
Thank you very much for your quick response and your interest in helping me. I appreciate it very much.

I explain in more detail the problem I am trying to solve:

First of all, I have been using Spacy version 2.1.3, during the code I have developed (Python) I need to do the following:

```
for token in doc:
    print(token.lemma_)
```

For Spacy version 2.1.3 this works correctly. However for Spacy version 3 this stops working. According to what I have read in different information threads, now the lemmatizer becomes part of the Pipeline. So, I have decided to do the following: 

```
 try:
            # source NER from the same pipeline package as the last component.
            self.nlp.add_pipe("ner", source=self.spacy.load("en_core_web_sm"))
            # insert the entity ruler
            self.nlp.add_pipe("entity_ruler", before="ner")
            # removing 'lemmatizer
            self.nlp.remove_pipe("lemmatizer")
            # Adding a new component in the pipeline: 'lemmatizer'.
            config={"model": None, "mode": "rule", "overwrite": False}
            self.lemmatizer = self.nlp.add_pipe("lemmatizer", config=config)
            # Initializing lemmatizer
            self.lemmatizer.initialize()
        except Exception as e:
            print(f "Component lemmatizer error: {e}")
```
As it is commented in the documentation, the NER component must be placed at the end and then I must import the lemmatizer component in a rule configuration mode. Anyway, my code for spacy version 3 does not work anymore.

Following the Lemmatizer GitHub code I have been using the function rule_lemmatize(token) but it does not correctly return the value I expect. For example, the token "fails" I want to be converted to "fail" but I can't get it. Also it returns the following message:

```
UserWarning: [W108] The rule-based lemmatizer did not find 
POS annotation for the token 'fails'. 
Check that your pipeline includes components that assign 
token.pos, typically 'tagger'+'attribute_ruler' or 'morphologizer'.
warnings.warn(Warnings.W108.format(text=string))
```

I would appreciate it if you could clarify how to solve this Spacy version update. If further explanation of the error is needed, I am open to it.

**Reply by polm** [reply]
OK, thanks for the extra information, that helps.

Are you using only the pretrained spaCy pipelines? Or do you have a custom NER model or something?

Since you are already using the pretrained English pipeline, it might be easiest for you to load that and keep the default lemmatizer and then add your custom components back. In your example code you are removing a lemmatizer in your pipeline and I'm not sure why that's necessary.

Example code could look something like this:

```
import spacy

# this default model includes a working lemmatizer
nlp = spacy.load("en_core_web_sm")

# this is your pipeline with custom components, if any
mynlp = spacy.load("mynlp")

# use this to copy your custom components
nlp.add_pipe("xxx", source=mynlp)

ruler = nlp.add_pipe("entity_ruler", before="ner")

# now nlp contains your custom components, entity ruler, and lemmatizer
```

**Reply by joaneeet7** [reply]
An important fact is the following:

When I extract the doc object I do it by using the tokenizer I created previously. 

`doc = self.tokenizer(text)`

It may be a piece of information to take into account as well. I'm not sure if that affects the problem I have.

**Reply by polm** [reply]
Uh, OK. Why are you calling the tokenizer directly rather than using a Language object (like `nlp` in the example)?

**Reply by joaneeet7** [reply]
I have a custom tokenizer.

**Reply by adrianeboyd** [reply]
Your v2 code may have been using the backoff lookup lemmatizer rather than the rule-based lemmatizer if your pipeline did not contain a `tagger`. If you want the same behavior in a v3 pipeline, see how to add the English lookup lemmatizer here under "Switch from rule-based to lookup lemmatization": https://spacy.io/models#design-modify

**Reply by joaneeet7** [reply]
Thanks for your suggestions. The intention I have is to use only the tokenizer to perform the token creation. However, the tokens do not have the lemma_ property if the tokenizer is used. I attach a code with the example results:

```
import spacy
nlp = spacy.load('en_core_web_sm')
text = 'apt installation of kubeadm fails: Depends: kubelet (>= 1.6.0) but it is not going to be installed'.
# With NLP
doc_nlp = nlp(text)
for token in doc_nlp:
    print(token.lemma_)
# ======================================
tokenizer= spacy.tokenizer.Tokenizer(
    nlp.vocab,
    rules=spacy.lang.en.tokenizer_exceptions.TOKENIZER_EXCEPTIONS
)
# With Tokenizer
doc_tokenizer = tokenizer(text)
for token in doc_tokenizer:
    print(token.lemma_)
```


When done with tokenizer I can't extract the lemma from the token. I am interested in doing the process with the tokenizer because nlp is 3 times slower.

**Reply by adrianeboyd** [reply]
The lookup lemmatizer worked differently in v2. In v3 you have to explicitly add a `lemmatizer` to your pipeline. The English default mode is `rule` and requires `tagger`+`attribute_ruler`, but the `lookup` lemmatizer does not and should be fast.

**Reply by polm** [reply]
Also note that if you do this:

```
nlp = spacy.load('en_core_web_sm')
``` 

You're going to be running components like the parser that you don't need, and turning them off will also help. See the speed FAQ:  #8402

**Reply by joaneeet7** [reply]
Thanks for your support!

I have changed the code and I'm using NLP to do the tokenization.

**Reply by joaneeet7** [reply]
I have had a small problem with the issue of extracting a lemma. For example the word installed is passed to the word instal (with one 'l'), I do not understand why it does not return install (with two 'l'). Could you explain to me the reason for changing the word 'installed' to 'instal'?

**Reply by polm** [reply]
"instal" is [a variant spelling](https://www.merriam-webster.com/dictionary/instal) and happens to be the first lemma listed (probably for sort order reasons). The lemma data is from WordNet originally, you can check there if you're curious about something.
